### PR TITLE
Add integration of TT-TDB; add and update GHAs

### PR DESCRIPTION
**Pull request by PerezHz**
This PR adds to the `DE430!` dynamical model the integration of TT-TDB following essentially the approach of [Fienga et al. (2009)](https://ui.adsabs.harvard.edu/abs/2009A%26A...507.1675F/abstract), while adding the contribution to TT-TDB due to the Sun's J_2 as is done in the DE430/431 integration.

cc: @lbenet @LuEdRaMo

### CODE REVIEW

**Comment by PerezHz** [comment]
P.S. I can rebase this PR after merging #16, if the latter is ready to be merged

**Comment by PerezHz** [comment]
The core of the changes of this PR is at the end of `src/dynamical_model.jl`, with corresponding changes in `src/jetcoeffs.jl`; the rest of the changes are either aesthetic or minor updates.

**Comment by PerezHz** [comment]
(rebased to current `main`)

**Comment by PerezHz** [comment]
Since it's a small and independent change, I'm also adding here CI and CompatHelper for PlanetaryEphemeris just to avoid having to open another PR; otherwise this is ready for review

**Comment by PerezHz** [comment]
Update: tests are showing that `datetime2julian` tests for `Float128` are failing since, AFAIU, the corresponding methods were removed in #16 to avoid type piracies. I temporarily commented out those tests to see how CI reacts, but since the only allowed initial conditions are the ones in 1969 and 2000, I would think that those tests are maybe not longer needed? How do you suggest @lbenet @LuEdRaMo we should handle this?

**Comment by lbenet** [comment]
> Update: tests are showing that `datetime2julian` tests for `Float128` are failing since, AFAIU, the corresponding methods were removed in #16 to avoid type piracies. I temporarily commented out those tests to see how CI reacts, but since the only allowed initial conditions are the ones in 1969 and 2000, I would think that those tests are maybe not longer needed? How do you suggest @lbenet @LuEdRaMo we should handle this?

For the time being, I think it is ok to have them commented. If we want those type of checks, I think a PR should be opened in Dates...

File: src/jetcoeffs.jl [hunk 1]
```diff
@@ -1,14 +1,16 @@
+### LASCIATE OGNE SPERANZA, VOI CH'INTRATE ###
```
Comment (lbenet) [inline src/jetcoeffs.jl:1]: Include an english translation, it's worth! 😄

**Review by lbenet — COMMENTED** [review]
Thanks @PerezHz for this *needed* addition! I've left few comments, which may improve in terms of performance... or may be trivially not useful. 

In any case, the question is if we can test the results (say, against the JPL)? (While those tests are not needed, having a script may be useful for long-term peace-of-mind...

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -40,4 +40,4 @@ jobs:
           files: lcov.info
       - uses: julia-actions/julia-uploadcoveralls@v1
         env:
-          COVERALLS_TOKEN: ${{ secrets.COVERALLS_TOKEN }}
+          COVERALLS_TOKEN: ${{ secrets.COVERALLS_REPO_TOKEN }}
```
Comment (lbenet) [inline .github/workflows/CI.yml:43]: It seems this change solve the issues, right?

**Comment by LuEdRaMo** [comment]
I think I have some better tests and a few improvements to the `selecteph` function. Should I push those changes here or leave it for another PR?

**Comment by PerezHz** [comment]
> I think I have some better tests and a few improvements to the selecteph function. Should I push those changes here or leave it for another PR?

It will take me a couple of days to address all of @lbenet's review points (hopefully it'll be ready by tomorrow!), so if your tests are already on top of current main branch, maybe it makes sense to open another PR; then we can rebase here if needed

File: .github/workflows/CI.yml (hunk 2 above)
Comment (PerezHz) [reply .github/workflows/CI.yml:43]: Indeed!

File: src/dynamical_model.jl [hunk 3]
```diff
@@ -3148,8 +3151,41 @@ function DE430!(dq, q, params, t)
     dq[6N+12] = inv_I_c_t[3,3]*Ic_dωc_3 # + ( (inv_I_c_t[3,1]*Ic_dωc_1) + (inv_I_c_t[3,2]*Ic_dωc_2) )
 
     # TT-TDB
-    # TODO: implement TT-TDB integration
-    dq[6N+13] = zero_q_1
+
+    # Contribution to TT-TDB due to Solar J2
+    # See equation (7) in page 7 of https://ui.adsabs.harvard.edu/abs/2014IPNPR.196C...1F%2F/abstract
+    w_LE = (newtonianCoeff[su,ea]*J2_t[su])*((one_t - (3(sin_ϕ[su,ea]^2)))/2)
```
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/dynamical_model.jl:3157]: > Does J2_t[su] include the squared solar radius? (Just being sure...)

Yes, it's in `J2_t` (I agree the name is somewhat misleading; probably it makes sense to rename this variable)

File: src/dynamical_model.jl [hunk 4]
```diff
@@ -3148,8 +3151,41 @@ function DE430!(dq, q, params, t)
...
+        else
+            β_TTmTDB_i_1 = 4( ((dq[3i-2]*dq[3ea-2])+(dq[3i-1]*dq[3ea-1])) + (dq[3i]*dq[3ea]) )
+            β_TTmTDB_i_2 = newtonianNb_Potential[i] - ( (1.5v2[ea]) + (2v2[i]) )
+            β_TTmTDB_i_3 = (  ((dq[3(N+i)-2]*X[i,ea]) + (dq[3(N+i)-1]*Y[i,ea])) + (dq[3(N+i)]*Z[i,ea])  ) /2
+            vi_dot_rEi = ((dq[3i-2]*X[ea,i]) + (dq[3i-1]*Y[ea,i])) + (dq[3i]*Z[ea,i])
+            vi_dot_rEi_div_rEi_sq = (vi_dot_rEi/r_p1d2[ea,i])^2
+            β_TTmTDB_i_4 = vi_dot_rEi_div_rEi_sq/2
+            β_TTmTDB_i = ((β_TTmTDB_i_1 + β_TTmTDB_i_2) + ( β_TTmTDB_i_3 + β_TTmTDB_i_4))
+            temp_β_TTmTDB = β_TTmTDB + (newtonian1b_Potential[i,ea]*β_TTmTDB_i)
+            β_TTmTDB = temp_β_TTmTDB
```
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/dynamical_model.jl:3180]: That's a good point, I'll address this in the next few commits

File: src/dynamical_model.jl [hunk 5]
```diff
@@ -3148,8 +3151,41 @@ function DE430!(dq, q, params, t)
...
+        end # else (i != j)
+    end
+
+    # d(TT-TDB)/d(TDB)
+    # See equation (10) of https://ui.adsabs.harvard.edu/abs/2009A%26A...507.1675F/abstract
+    # Since the initial condition is in seconds, and the independent variable is in days,
+    # an additional factor daysec=86400 is needed for unit consistency
+    dq[6N+13] = daysec*(( L_B_t - c_m2*α_TTmTDB)*(one_t + L_B_t - L_G_t) + ( (c_m4*β_TTmTDB) - L_G_t ))
```
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/dynamical_model.jl:3188]: Agreed 👍

**Comment by LuEdRaMo** [comment]
> It will take me a couple of days to address all of @lbenet's review points (hopefully it'll be ready by tomorrow!), so if your tests are already on top of current main branch, maybe it makes sense to open another PR; then we can rebase here if needed

Actually the tests are on the top of this PR.

File: test/propagation.jl [hunk 6]
```diff
@@ -1,43 +1,42 @@
...
-using Dates 
+using Dates
 using Quadmath
 using TaylorIntegration
-using TaylorSeries 
+using TaylorSeries
 using JLD2
 using Test
 
 using PlanetaryEphemeris: initialcond
```
[the remark this answers is not in the document]
Comment (lbenet) [reply test/propagation.jl:11]: I forgot to say in the previous comment that I was referring to the Project.toml file... It is also worth to upgrade TaylorIntegration version there (to 0.13).

**Comment by lbenet** [comment]
Just a trivial observation from the last tests run: v1.6 takes 31m to complete the tests, while v1.9 19m, and nightly ~10!

**Comment by LuEdRaMo** [comment]
Just as a reference, after 100 steps (roughly 228.5 days), the difference in TT-TDB between PE and JPL is approximately 1.56e-10 seconds.

**Comment by PerezHz** [comment]
> Actually the tests are on the top of this PR.

Alright, then it's just easier to push here; feel free to push to this PR as soon as you're ready 😄

**Comment by PerezHz** [comment]
> Just as a reference, after 100 steps (roughly 228.5 days), the difference in TT-TDB between PE and JPL is approximately 1.56e-10 seconds.

0.1 ns in ~1 yr is not too bad... Did your run include all the asteroids?

**Comment by LuEdRaMo** [comment]
> 0.1 ns in ~1 yr is not too bad... Did your run include all the asteroids?

Yes. I will do a 100 years integration to see the long-term behaviour, but that will take some time.

File: test/propagation.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (PerezHz) [reply test/propagation.jl]: Ok, I'll make sure we tackle that over PerezHz/NEOs.jl#11

**Comment by LuEdRaMo** [comment]
Tests seem to work but they are slow. Is there a way you know to run them on multiple threads?

**Comment by lbenet** [comment]
I think there is a way to use more threads in the tests, but I'm not so sure if that would be the best with respect to Github...

**Comment by lbenet** [comment]
> Tests seem to work but they are slow.

Could it be the precompilation step?

**Comment by PerezHz** [comment]
> Tests seem to work but they are slow. Is there a way you know to run them on multiple threads?

There is indeed a way, by passing the `JULIA_NUM_THREADS` environment variable to the test run step in the CI worfklow, which I've now added to CI.yml file (I think there is a limit to use only two threads). I actually think it makes sense to perform our tests using two threads, if not for the sake of performance itself, to actually test the propagation in a multi-threaded environment.

**Comment by PerezHz** [comment]
> Could it be the precompilation step?

I think it is

**Comment by PerezHz** [comment]
> Could it be the precompilation step?

I think the longer times are (maybe?) also due to the fact that the tests also propagate using the `Float128` methods, so we are in a sense precompiling the specialized `jetcoeffs!` & co. methods twice (one for `Float64`, one for `Float128`)

**Comment by PerezHz** [comment]
> I think we should only compile Float64, which is what we currently use mostly. Also, percompiling few methods in TaylorSeries and TaylorIntegration should make precopilation easier here.

I agree with that; I pushed a commit to see how precompilation goes with `Float128`, but maybe we can settle for the time being on precompilation and tests only for `Float64`? And I also agree that precompiling stuff on TS and TI will help us here

File: src/dynamical_model.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/dynamical_model.jl]: This is done now

File: src/dynamical_model.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/dynamical_model.jl]: Done ✅

File: src/dynamical_model.jl [hunk 7]
```diff
@@ -2150,11 +2150,15 @@ function DE430!(dq, q, params, t)
...
     pntempZ = Array{S}(undef, N)        # Z-axis component
-    # Full post-Newtonian accelerations 
+    # Full post-Newtonian accelerations
     postNewtonX = Array{S}(undef, N)    # X-axis component
     postNewtonY = Array{S}(undef, N)    # Y-axis component
     postNewtonZ = Array{S}(undef, N)    # Z-axis component
 
+    ### TT-TDB computation
+    local L_G_t = L_G*one_t
+    local L_B_t = L_B*one_t
```
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/dynamical_model.jl]: Done ✅

**Comment by lbenet** [comment]
Tests pass now, except in Julia v1.6, because the tests were cancelled; I suggest **not** to test that version of Julia.

**Comment by PerezHz** [comment]
> Tests pass now, except in Julia v1.6, because the tests were cancelled; I suggest not to test that version of Julia.

Indeed, there seems to be a 2 hour timeout for CI workflows which the testing on julia1.6 exceeds; I modified CI.yml to test only 1.9 and nightly for now. This still leaves open the question of what to do with JuliaRegistrator: there is a 1 hour timeout for autoMerge tasks when registering a package in the General registry, and  since loading the package for the first time (i.e., doing `import PlanetaryEphemeris`) with precompilation enabled takes currently more than an hour there, we haven't been able to automatically register PlanetaryEphemeris v0.5.1 and something similar might happen with newer releases as well: one of the requirements for autoMerge is that packages should be loadable, but if loading takes more than an hour then the job is not finished and autoMerge fails. I actually think we maybe should comment out `precompile.jl` from the includes until we improve TTFL for PlanetaryEphemeris, what do you think?

**Comment by lbenet** [comment]
I was thinking the same: simply comment the precompilation step, and see how that behaves...

**Comment by LuEdRaMo** [comment]
I added a commit to try commenting out the precompilation file and doing only the Float64 tests.

**Comment by PerezHz** [comment]
> I was thinking the same: simply comment the precompilation step, and see how that behaves...

Agreed, and thanks @LuEdRaMo for pushing the corresponding changes!

As a side note, maybe we can consider somehow taking advantage of PackageCompiler.jl to tackle the issues here with TTFL. I'm thinking out loud here, but maybe we could build a system image containing the compiled `jetcoeffs!` & co. methods defined by PlanetaryEphemeris, and host that sysimg it somewhere and then just retrieve it when needed... For me it makes sense, since the dynamical model is not so frequently updated, so we would only re-build it every now and then.

Another option can be to put the dynamical model methods as a package extension, and load it as a weakdep, thus it doesn't get in the way of loading/compiling times. Of course, this latter option only delays/postpones the precompilation of the `jetcoeffs!` method until it's used... but no need to decide anything now; just putting some ideas around 😄

**Comment by PerezHz** [comment]
I just pushed a commit fixing a small issue with the tests, and updated the test env setup via a `test/Project.toml`

**Comment by PerezHz** [comment]
... and it went down to 32 minutes on 1.9 and nightly! I think we're in business, folks!

**Comment by PerezHz** [comment]
I'm currently adding some comparisons with the JPL ephemeris in the tests; hope to have that ready ideally by today...

**Comment by lbenet** [comment]
This this is almost ready, modulo the tests you are adding.

The question is how shall we proceed now? Merge this and tag the corresponding version as 0.6.0 (and simply forget about 0.5.1)?

**Comment by PerezHz** [comment]
> The question is how shall we proceed now? Merge this and tag the corresponding version as 0.6.0 (and simply forget about 0.5.1)?

I agree going with v0.6.0; just updated the Project.toml accordingly. If tests pass then this PR will be ready to be merged.

**Comment by PerezHz** [comment]
Tests are passing so I'll go ahead and merge.

**Pull request merged**